keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,642 @@
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The Probe struct contains fields that specify seconds for intervals.
+Some users would like to have intervals less than or slightly greater
+than one second.
+
+## Motivation
+
+Better performance in the form of faster startups, more timely readiness checks...
+
+#### Knative
+
+For example, Knative will create Pods (via Deployment) and wait for them to become `Ready`
+when handling an HTTP request from an end-user. In this case, Pod readiness
+latency has a direct impact on HTTP request latency. (In the steady state,
+Knative will re-use an existing Pod, but this situation can happen on scale-up,
+and is guaranteed to happen on scale-from-zero.)
+
+### Goals
+
+An ability to specify probe durations that have more granularity, more specificity, than just zero,
+one, or n-seconds. Focus is on possibly faster readiness probes, for example every .5 seconds until succss
+and more granular use beyond existing defaults, for example 1.5 second duration instead of picking
+between 1 or 2 second duration.
+Add additional tests cases to the timeout test cases.
+Not breaking backwards compatibility with the V1 API.
+
+
+### Non-Goals
+
+V2 API for existing objects.
+Converting fields from `int32` to `resource.Quantity`.
+
+
+## Proposal
+
+TLDR:
+Add two new optional fields (of type `*int32`) to the Probe struct, which would be used to offset the second based time values in the Probe struct:
+
+- `PeriodMilliseconds`       // How often (in milliseconds) to offset PeriodSeconds when performing the probe (*** to reduce un-necessary resource usage, when periodMilliseconds is used to reduce the period to less than a second the offset is only used until success is reached then no longer).
```

Comment:
This is somewhat unusual - the "period offset used the first time".  While it is the most backwards compatible version of the change, it is very non obvious to users and potentially non-obvious to implementers from the field name.  I notice in your comments that we are tending to use the word "offset" to describe this field - it would be clearer if the field name included offset as well - perhaps `periodOffsetMilliseconds` and `initialDelayOffsetMilliseconds`.

Reply:
Also, the godoc should include the range of allowed values - `[-999,999]` if that is intended (as a later comment implied), as well as other implications.  Might as well get these in shape so that the proposal more clearly summarizes the change.

Reply:
Finally, the comment about unnecessary resource usage is confusing and doesn't seem to be justified in the proposal below (I see comments about more data needed?). You might want to reword and summarize this (in what we show to users) as

> Probe periods shorter than one second may be lengthened when a node is under resource pressure.

In general I think it's well within the bounds of a health checker implementation to probe less often than requested, which the proposal should clarify even if we are vague about when that may happen to the user.  While we could be very specific (perhaps the 200 and 500ms values from below), it's not clear from a conformance perspective how we would enforce this value (or even if we should).

Reply:
nod..

Reply:
incorporated..